### PR TITLE
Implement RFC 1234: Ignore PhantomData when checking CoerceUnsized implementations

### LINKED ISSUE #26905: DST coercions are prevented by phantom data
This prevents Rc from having a sound definition with respect to dropcheck

```rust
#![feature(unsize, coerce_unsized)]

use std::marker::{Unsize, PhantomData};
use std::ops::CoerceUnsized;

struct MyRc<T: ?Sized> {
    ptr: *const T,
    _boo: PhantomData<T>,
}

impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<MyRc<U>> for MyRc<T>{ }

fn main() {}
```

```text
<anon>:11:1: 11:77 error: the trait `CoerceUnsized` may only be implemented for a coercion between structures with one field being coerced, but 2 fields need coercions: ptr (*const T to *const U), _boo (core::marker::PhantomData<T> to core::marker::PhantomData<U>) [E0375]
<anon>:11 impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<MyRc<U>> for MyRc<T>{ }
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
```

### LINKED ISSUE #28239: Tracking issue for PhantomData being permitted by CoerceUnsize (RFC #1234)
Tracking issue for https://github.com/rust-lang/rfcs/pull/1234

cc @Gankro

### PR DESCRIPTION
**Pull request by ebfull**
#27483 redux at Gankro's request.

Fixes #26905, Closes #28239

r? @nrc

### CODE REVIEW

File: src/librustc/middle/ty.rs [hunk 1]
```diff
@@ -4373,6 +4373,13 @@ impl<'tcx> TyS<'tcx> {
         }
     }
 
+    pub fn is_phantom_data(&self, cx: &ctxt<'tcx>) -> bool {
+        if let &TyStruct(def_id, _) = &self.sty {
+            return Some(def_id.did) == cx.lang_items.phantom_data();
```
Comment (arielb1) [inline src/librustc/middle/ty.rs]: `def.is_phantom_data()`

**Comment by ebfull** [comment]
@nrc This was rebased.

**Pull request merged**